circuits/cpp/barretenberg/cpp/src/barretenberg/proof_system/circuit_builder/eccvm/eccvm_circuit_builder.hpp

```diff
@@ -0,0 +1,524 @@
+                    .base_point = base_point,
+                    .wnaf_slices = compute_wnaf_slices(scalar),
+                    .wnaf_skew = (scalar & 1) == 0,
+                    .precomputed_table = compute_precomputed_table(base_point),
+                });
+                pc--;
+            }
+        };
+
+        for (auto& op : vm_operations) {
+            if (op.mul) {
+                process_mul(op.z1, op.base_point);
+                process_mul(op.z2, AffineElement{ op.base_point.x * FF::cube_root_of_unity(), -op.base_point.y });
+
+            } else {
+                if (!active_msm.empty()) {
+                    msms.push_back(active_msm);
+                    active_msm = {};
+                }
+            }
+        }
+        if (!active_msm.empty()) {
+            msms.push_back(active_msm);
+        }
+
+        ASSERT(pc == 0);
+        return msms;
+    }
+
+    static std::vector<ScalarMul> get_flattened_scalar_muls(const std::vector<MSM>& msms)
+    {
+        std::vector<ScalarMul> result;
+        for (const auto& msm : msms) {
+            for (const auto& mul : msm) {
+                result.push_back(mul);
+            }
+        }
+        return result;
+    }
+
+    void add_accumulate(const AffineElement& to_add)
+    {
+        vm_operations.emplace_back(VMOperation{
+            .add = true,
+            .mul = false,
+            .eq = false,
+            .reset = false,
+            .base_point = to_add,
+            .z1 = 0,
+            .z2 = 0,
+            .mul_scalar_full = 0,
+        });
+    }
+
+    void mul_accumulate(const AffineElement& to_mul, const CycleScalar& scalar)
+    {
+        CycleScalar z1 = 0;
+        CycleScalar z2 = 0;
+        auto converted = scalar.from_montgomery_form();
+        CycleScalar::split_into_endomorphism_scalars(converted, z1, z2);
+        z1 = z1.to_montgomery_form();
+        z2 = z2.to_montgomery_form();
+        vm_operations.emplace_back(VMOperation{
+            .add = false,
+            .mul = true,
+            .eq = false,
+            .reset = false,
+            .base_point = to_mul,
+            .z1 = z1,
+            .z2 = z2,
+            .mul_scalar_full = scalar,
+        });
+    }
+    void eq(const AffineElement& expected)
+    {
+        vm_operations.emplace_back(VMOperation{
+            .add = false,
+            .mul = false,
+            .eq = true,
+            .reset = true,
+            .base_point = expected,
+            .z1 = 0,
+            .z2 = 0,
+            .mul_scalar_full = 0,
+        });
+    }
+
+    void empty_row()
+    {
+        vm_operations.emplace_back(VMOperation{
+            .add = false,
+            .mul = false,
+            .eq = false,
+            .reset = false,
+            .base_point = CycleGroup::affine_point_at_infinity,
+            .z1 = 0,
+            .z2 = 0,
+            .mul_scalar_full = 0,
+        });
+    }
+
+    RawPolynomials compute_full_polynomials()
```

Comment:
Since this is a core function that actually returns all the date needed for the prover, it would be great if there was an explanation of all the members and what they should contain, at least loosely. Right now a person would have to cross-reference with several other files while trying to understand what actually goes into the final polynomials

Reply:
will do